temporal-sdk/src/main/java/io/temporal/client/ListWorkflowExecutionIterator.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.client;
+
+import com.google.protobuf.ByteString;
+import io.temporal.api.workflow.v1.WorkflowExecutionInfo;
+import io.temporal.api.workflowservice.v1.ListWorkflowExecutionsRequest;
+import io.temporal.api.workflowservice.v1.ListWorkflowExecutionsResponse;
+import io.temporal.internal.client.external.GenericWorkflowClient;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ListWorkflowExecutionIterator implements Iterator<WorkflowExecutionInfo> {
+  private static final Logger log = LoggerFactory.getLogger(ListWorkflowExecutionIterator.class);
+
+  private final @Nonnull String query;
+  private final @Nonnull String namespace;
+  private final int pageSize;
+  private final @Nonnull GenericWorkflowClient genericClient;
+  private ListWorkflowExecutionsResponse activeResponse;
+  private int nextActiveResponseIndex;
+  private CompletableFuture<ListWorkflowExecutionsResponse> nextResponse;
+
+  public ListWorkflowExecutionIterator(
+      @Nonnull String query,
+      @Nonnull String namespace,
+      int pageSize,
+      @Nonnull GenericWorkflowClient genericClient) {
```

Comment:
Since this takes an internal type, could justify `ListWorkflowExecutionIterator` only being an interface and moving this impl back to `io.temporal.internal.client` or `io.temporal.internal.client.external` somewhere. But this works for me as is.